components/dispute-game-contracts/README.md

```diff
@@ -0,0 +1,276 @@
+
+# Context
+
+We assume that the sequencer committed to some transition in the roll-up's state $T = (R, transactions) \to R'$, where R' is the state root (aka `chainState.root`) after applying `transactions` to R, in order.
+
+Suppose a challenger wants to challenge this state. Consider a `MachineState` type something like the following:
+
+```tsx
+type MachineState {
+  root: Bytes32; // hash of (chainState.root, memory.root, instructions)
+  chainState: ChainState;
+  memory: Memory;
+  instructions: Instruction;
+  // etc
+}
+```
+
+(MachineState is assumed to be a Merkle tree, thus the `root` property.)
+
+We force the sequencer to post `S_0.root`, where `S_0 = {chainState: R, memory: emptyMemory, instructions: dissassemble(transactions)}` . The challenger contract must validate that `R` is equal to `R` posted for normal operation rollup. In addition, the challenger contract must validate that `dissassemble(transaction)` is correct. These validations are currently out of scope. We also force the sequencer to post `S'.root`, where `S' = {chainState: R', memory: emptyMemory, instructions: emptyInstructions}` . We might force the sequencer to post additional intermediate states.
+
+We assume there is a function `progress: (state: MachineState) => MachineState` which takes in a machine state `s` and returns the next state. It works by popping the latest instruction, and modifying itself according to the instruction to produce the next state. When `instructions` is empty, `progress(state)` should indicate that it's reached a final state.
+
+The following description is restricted in scope to _correctness_ requirements, and ignores _liveness_ requirements, **assuming that the sequencer & challenger each act in a timely manner.** The specific incentives & timeouts are not considered in this document.
+
+# High level overview
+
+The `progress` function _defines_ what it means for a transition $(R, transactions) \to R'$ to be valid, in the following sense:
+
+1. From $(R, transactions)$, one can construct the "initial" machine state $M_0$.
+2. Define $M_{i+1} = progress(M_i)$
+3. Compute $M_1, M_2, ...$ until you arrive at a "terminal" state $M_t$
+4. Check that $M_t.chainState.root == R'$. If yes, $(R, transactions) \to R'$ is valid; else, it is invalid.
+
+In the event of a validator disputing a sequencer-posted state root, the "bisection game" plays out as follows.
+
+1. [Off-chain] The sequencer computes _its_ belief about what the correct sequence of machine states is, generating a sequence $S_0, ..., S_n$. (A correctly behaving sequencer would find $M_i = S_i$.)
+2. [Off-chain] The validator computes _their_ belief about what the correct sequence of machine states is, generating a sequence $V_0, ..., V_m$. (A correctly behaving validator would find $M_i = V_i$.)
+3. [On-chain] The sequencer commits to the root of the machine state at some steps $step_0, step_1, ..., step_k$, where initially, $step_k = \lfloor k/n \rfloor$.
```

Comment:
Sadly, github doesn't render latex. I thought it might, since VSCode seems to without any extra configuration:
<img width="1531" alt="Screen Shot 2021-08-27 at 5 04 51 PM" src="https://user-images.githubusercontent.com/8432675/131199534-7c440f7f-df4d-4967-9077-088e372e2840.png">